Feature: environments (#41)
giantswarm/gitops-template: docs/add_wc_environments.md

- [Environments](#environments)
  - [Stages](#stages)
    - [The development cluster](#the-development-cluster)
    - [The staging cluster](#the-staging-cluster)
    - [The production cluster](#the-production-cluster)
- [Add Workload Clusters based on the environment cluster templates](#add-workload-clusters-based-on-the-environment-cluster-templates)
- [Tips for developing environments](#tips-for-developing-environments)

You might want to set up multiple, similar Workload Clusters that serve as for example development,
staging and production environments. You can utilize [bases](/bases) to achieve that. Let's take a look at the
[/bases/environments](/bases/environments) folder structure.

## General note

It is possible to solve the environment and environment propagation problem in multiple ways, notably by:

- using a multi-directory structure, where each environment is represented as a directory in a `main` branch of a single
  repository
- using a multi-branch approach, where each branch corresponds to one environment, but they are in the same repo
- using a multi-repo setup, where there's one root repository providing all the necessary templates, then there is one another

>   repository per each environment.

Reviewer: ```suggestion
  repository per environment.
```

> Each of these approaches has pros and cons. Over here, we're proposing the multi-directory approach. The pros of it are:

Reviewer: ```suggestion
Each of these approaches has pros and cons. We propose the multi-directory approach. The pros of it are:
```

a single repo and branch serving as the source of truth for all the environments, very easy template sharing and
relatively easy way to compare and promote configuration across environments. On the other hand, it might not be the

> best solution for access control, templates versioning and also easy comparing of environments.

Reviewer: ```suggestion
best solution for access control, template versioning and also easy comparing of environments.
```

differentating factor for that kind of environment, then you should create subfolder there for different possible values.
For example, for multiple data centers, we recommend putting region specific configuration into
`/bases/environments/regions` folder and under there create `ap-east-1`, `eu-central-1` and `us-west-2` folders.

Once your environment templates are ready, you can use them to create new clusters by placing cluster definitions in
in [/management-clusters/MC_NAME/organizations/ORG_NAME/workload-clusters](
/management-clusters/MC_NAME/organizations/ORG_NAME/workload-clusters)

> :construction: Please note that if you want to use multiple environment templates to create a single cluster
that uses `App CR`s for deployments, like you would like to use `dev` out of `staging` layout to set app configuration
and then use `east` from the `data-centers` to set the IP ranges, you will run into issues around merging
configurations, as currently one configuration source (ie. `ConfigMap` in `spec.config.configMap`) completely
overrides the whole value of the same attribute coming from the other base. We're working to remove this limitation.

### Stages

> We have 3 example clusters under [/bases/environments/stages](/bases/environments/stages):

Reviewer: All other docs follow an instruction format of "this is what you need to do to have environments". Here we now have a mix of "look at these files & this is what you can do". Should there be a clear guide of "this is how you add a stage"?

Reviewer 2: I think it would make sense and I would love it. I was pondering that myself as well while writing docs.

Reviewer 2: @MarcelMue Created a ticket for this to discuss async. Also addes `topic/documentation` as it might make sense for `sig/docs` as well.
